cpp/src/neighbors/detail/cagra/cagra_build.cuh

```diff
@@ -155,32 +230,59 @@ void build_knn_graph(
 
     cuvs::neighbors::ivf_pq::search(
       res, *search_params, index, queries_view, neighbors_view, distances_view);
-    if constexpr (raft::is_host_mdspan_v<decltype(dataset)>) {
+
+    if (!device_refinement) {
```

Comment:
For managed arrays (expected to be large, and residing in host memory) we should use host refinement.